src/Functions/GatherUtils/Algorithms.h

```diff
@@ -418,15 +433,93 @@ bool sliceHasImpl(const FirstSliceType & first, const SecondSliceType & second,
                 has = true;
         }
 
-        if (has && !all)
+        if (has && search_type == ArraySearchType::Any)
             return true;
 
-        if (!has && all)
+        if (!has && search_type == ArraySearchType::All)
             return false;
+    }
+    return search_type == ArraySearchType::All;
+}
+
+template < typename FirstSliceType,
+           typename SecondSliceType,
+           bool (*isEqual)(const FirstSliceType &, const SecondSliceType &, size_t, size_t),
+           bool (*isEqualUnary)(const SecondSliceType &, size_t, size_t)>
+bool sliceHasImplSubStr(const FirstSliceType & first, const SecondSliceType & second, const UInt8 * first_null_map, const UInt8 * second_null_map) noexcept
+{
+    const bool has_first_null_map = first_null_map != nullptr;
+    const bool has_second_null_map = second_null_map != nullptr;
+
+    if (second.size == 0) 
+        return true;
+
+    const auto aux = buildKMPFailureArray<SecondSliceType, isEqualUnary>(second, second_null_map);
 
+    size_t secondCur = 0;
+    size_t firstCur = 0;
+
+    while (firstCur < first.size && secondCur < second.size)
+    {
+        auto cond_values_match = isEqual(first, second, firstCur, secondCur);
+
+        const bool is_first_null = has_first_null_map && first_null_map[firstCur];
+        const bool is_second_null = has_second_null_map && second_null_map[secondCur];
+
+
+        auto cond_null_match = is_first_null && is_second_null;
+        if (cond_values_match || cond_null_match)
+        {
+            ++firstCur;
+            ++secondCur;
+        }
+        else if (secondCur > 0)
+        {
+            secondCur = aux.get()[secondCur - 1];
+            }
+        else
+        {
+            ++firstCur;
+        }
     }
+    if (secondCur == second.size)
+    {
+        return true;
+    }
+    return false;
+}
 
-    return all;
+template <typename SliceType, bool (*isEqual)(const SliceType &, size_t, size_t)>
+std::unique_ptr<size_t[]> buildKMPFailureArray(const SliceType & pattern, const UInt8 * pattern_null_map) noexcept
+{
+    auto aux = std::make_unique<size_t[]>(pattern.size);
+    auto has_null_map = pattern_null_map != nullptr;
+    aux[0] = 0;
+    for (size_t i = 1; i < pattern.size; ++i)
+    {
+        auto length = aux[i - 1];
+        while (length > 0)
+        {
+            auto cond_values_match = isEqual(pattern, i, length);
+            auto cond_null_map_match = (has_null_map && pattern_null_map[i] == pattern_null_map[length]);
+            if (cond_values_match || cond_null_map_match)
+            {
+                aux[i] = length + 1;
+                break;
+            }
+            else
+            {
+                length = aux[length - 1];
+            }
+        }
+        auto cond_values_match = isEqual(pattern, i, 0);
+        auto cond_null_map_match = (has_null_map && pattern_null_map[i] == pattern_null_map[0]);
+        if (length == 0 && (cond_values_match || cond_null_map_match))
```

Comment:
I would also suggest to change this piece as well, considering our discussion above.

Reply:
Ok, I will review and make the change.

Reply:
Change done.